scripts/misc/download_hf_model.py

```diff
@@ -0,0 +1,248 @@
+"""Download functions for Hugging Face models.
+
+TODO(jerry): Add better logging throughout
+TODO(jerry): Track download metrics
+
+Copyright 2022 MosaicML LLM Foundry authors
+SPDX-License-Identifier: Apache-2.0
+"""
+import os
+import sys
+import time
+
+import argparse
+from typing import Optional
+from http import HTTPStatus
+
+from bs4 import BeautifulSoup
+import huggingface_hub as hf_hub
+import requests
+
+# Constants derived from https://github.com/huggingface/transformers/blob/main/src/transformers/utils/__init__.py
+PYTORCH_WEIGHTS_NAME = "pytorch_model.bin"
+PYTORCH_WEIGHTS_INDEX_NAME = "pytorch_model.bin.index.json"
+PYTORCH_WEIGHTS_PATTERN = "pytorch_model*.bin*"
+
+SAFETENSORS_WEIGHTS_NAME = "model.safetensors"
+SAFETENSORS_WEIGHTS_INDEX_NAME = "model.safetensors.index.json"
+SAFETENSORS_EXCLUDE_PATTERN = "model*.safetensors*"
+
+DEFAULT_HF_HUB_CACHE_DIR = os.getenv(
+    "HUGGINGFACE_HUB_CACHE", os.path.expanduser("~/.cache/huggingface/hub")
+)
+
+
+def download_from_hf_hub(
```

Comment:
lets move both of these functions into the main llmfoundry library code, and then write unit tests for them.

Reply:
Moved into `llmfoundry/utils/model_download_utils.py`. Unit testing is a little tricky though - not sure if there's a clean way to test file downloads over network. I could probably use `pyfakefs` somehow, but I'm not sure how to easily mock the Hugging Face or cache server behavior

Reply:
i'd say don't worry about testing the network call itself, so just mock `snapshot_download` and `session.get`, to just test the code around the network call.